### PR TITLE
Can 358/add/filters on landing page

### PR DESCRIPTION
**Pull request by anders0204**
Added actual working filters on the landing page using bokeh widgets.

All the filters are handled by the `LandingPageFiltering` class which was nice at the start, but got a bit cluttered with 2+ filters. Could try to put the filters in another file.

One problem I see is that we don't have HPV18+ in the test data (only HPV18-).

Sorry for the large PR again 🫣

### CODE REVIEW

File: bokeh-demo/examples/landing_page/main.py [hunk 1]
```diff
@@ -1,25 +1,249 @@
...
+        start, end = birthyears.min(), birthyears.max()
+        age_slider = RangeSlider(
+            name="age_slider",
+            start=start,
+            end=end,
+            value=(start, end),
+            step=1,
+            title="Birth Year",
+        )
+        return age_slider
```
Comment (Northo) [inline bokeh-demo/examples/landing_page/main.py:106]:
```suggestion
        return RangeSlider(
            name="age_slider",
            start=start,
            end=end,
            value=(start, end),
            step=1,
            title="Birth Year",
        )
```

File: bokeh-demo/examples/landing_page/main.py [hunk 2]
```diff
@@ -1,25 +1,249 @@
...
+        age_min, age_max = self.age_slider.value
+
+        age_pids = list(
+            self.person_df[
+                self.person_df[self.column_names.birthdate].dt.year.between(
+                    age_min, age_max
+                )
+            ].index
+        )
+        return age_pids
```
Comment (Northo) [inline bokeh-demo/examples/landing_page/main.py:118]:
```suggestion
        return list(
            self.person_df[
                self.person_df[self.column_names.birthdate].dt.year.between(
                    age_min, age_max
                )
            ].index
        )
```

File: bokeh-demo/examples/landing_page/main.py [hunk 3]
```diff
@@ -1,25 +1,249 @@
...
+            "HR Histology",
+            "HPV16+",
+            "HPV18+",
+        ]
+
+        buttons = CheckboxButtonGroup(
+            name="checkbox_buttons",
+            labels=labels,
+        )
+        return buttons
```
Comment (Northo) [inline bokeh-demo/examples/landing_page/main.py:188]:
```suggestion
        return CheckboxButtonGroup(
            name="checkbox_buttons",
            labels=labels,
        )
```

File: bokeh-demo/examples/landing_page/main.py [hunk 4]
```diff
@@ -1,25 +1,249 @@
...
+    def _init_min_age_last_exam_slider(self) -> Slider:
+        min_age_last_exam_slider = Slider(
+            name="min_age_last_exam",
+            start=0,
+            end=100,
+            value=0,
+            step=1,
+            title="Minimum age at last exam",
+        )
+        return min_age_last_exam_slider
```
Comment (Northo) [inline bokeh-demo/examples/landing_page/main.py:206]:
```suggestion
        return Slider(
            name="min_age_last_exam",
            start=0,
            end=100,
            value=0,
            step=1,
            title="Minimum age at last exam",
        )
```

File: bokeh-demo/examples/landing_page/main.py [hunk 5]
```diff
@@ -1,25 +1,249 @@
...
+    def _init_min_n_screenings_slider(self) -> Slider:
+        min_n_screenings_slider = Slider(
+            name="min_n_screenings",
+            start=0,
+            end=10,
+            value=0,
+            step=1,
+            title="Minimum number of screenings",
+        )
+        return min_n_screenings_slider
```
Comment (Northo) [inline bokeh-demo/examples/landing_page/main.py:224]:
```suggestion
        return Slider(
            name="min_n_screenings",
            start=0,
            end=10,
            value=0,
            step=1,
            title="Minimum number of screenings",
        )
```

File: bokeh-demo/examples/landing_page/main.py (hunk 1 above)
Comment (Northo) [reply bokeh-demo/examples/landing_page/main.py:106]: According to https://docs.sourcery.ai/Reference/Python/Default-Rules/inline-immediately-returned-variable/. Up to you if you want this or not.

**Review by Northo — CHANGES_REQUESTED** [review]
Really nice, @anders0204 !

Added a couple of suggestions/comments, but mostly related to style preferences, so it's all up to you.

The things I think is the most important to change, is
- name of the age_slider, which I believe is more a year slider (or age of birth-slider)?
- the if 0 in active_buttons thing
- using logger instead of print (simply because it makes life easier later when running in container on tsd)

Some suggested tests
```python
# tests/test_landing_page.py
import pytest
import json
from datetime import datetime
import itertools

from examples.landing_page.main import LandingPageFiltering


@pytest.fixture
def landing_page_filtering(tmp_path, monkeypatch):
    selected_pids_path = tmp_path / "selected_pids.json"
    monkeypatch.setattr(
        "examples.landing_page.main.settings.selected_pids_path", selected_pids_path
    )
    landing_page_filtering = LandingPageFiltering()
    return landing_page_filtering, selected_pids_path


def test_landing_page_filtering_save(landing_page_filtering):
    landing_page_filtering, selected_pids_path = landing_page_filtering

    landing_page_filtering.save_pids()
    assert selected_pids_path.is_file()
    with open(selected_pids_path, "r") as f:
        selected_pids = json.load(f)
    assert selected_pids == landing_page_filtering.pid_list

def test_landing_page_filtering_min_age_last_exam_slider_pids(landing_page_filtering):
    landing_page_filtering, selected_pids_path = landing_page_filtering

    value = 50
    landing_page_filtering.min_age_last_exam_slider.value = value
    pids = landing_page_filtering.min_age_last_exam_slider_pids()
    assert len(pids) == 35

    last_exam_after_value = landing_page_filtering.person_df["age_last_exam"] >= value
    not_selected_pids = ~landing_page_filtering.person_df.index.isin(pids)

    assert all(last_exam_after_value[pids])
    assert not any(last_exam_after_value[not_selected_pids])

def test_landing_page_filtering_min_n_screenings_slider_pids(landing_page_filtering):
    landing_page_filtering, selected_pids_path = landing_page_filtering

    value = 3
    landing_page_filtering.min_n_screenings_slider.value = value
    pids = landing_page_filtering.min_n_screenings_slider_pids()
    assert len(pids) == 39

    n_screenings_after_value = landing_page_filtering.person_df["number_of_screenings"] >= value
    not_selected_pids = ~landing_page_filtering.person_df.index.isin(pids)

    assert all(n_screenings_after_value[pids])
    assert not any(n_screenings_after_value[not_selected_pids])

def test_landing_page_filtering_age_slider_pids(landing_page_filtering):
    landing_page_filtering, selected_pids_path = landing_page_filtering

    # NB! the age slider name is misleading, it does expect calendar years, not ages.

    min_date, max_date = datetime(1970, 1, 1), datetime(1997, 12, 31)
    landing_page_filtering.age_slider.value = (min_date.year, max_date.year)
    pids = landing_page_filtering.age_slider_pids()
    assert len(pids) == 34

    year_in_range = landing_page_filtering.person_df["FOEDT"].between(min_date, max_date)
    not_selected_pids = ~landing_page_filtering.person_df.index.isin(pids)

    assert all(year_in_range[pids])
    assert not any(year_in_range[not_selected_pids])


def test_landing_page_filtering_checkbox_button_pids(landing_page_filtering):
    landing_page_filtering, selected_pids_path = landing_page_filtering

    # Maybe a bit lazy, but just check that activating more buttons gives fewer pids.
    # Iterator that gives ((), (1,), (1,2), ...)
    active_buttons_iter = itertools.accumulate(
        range(len(landing_page_filtering.checkbox_buttons.labels)),
        lambda previous, new: (*previous, new),
        initial=(),
        )
    filtered_pids = []
    for active_buttons in active_buttons_iter:
        landing_page_filtering.checkbox_buttons.active = list(active_buttons)
        filtered_pids.append(landing_page_filtering.checkbox_button_pids())
    assert (a <= b for a, b in itertools.pairwise(filtered_pids))
```

File: bokeh-demo/examples/landing_page/main.py [hunk 6]
```diff
@@ -1,25 +1,249 @@
...
+
+    def _init_checkbox_buttons(self) -> CheckboxButtonGroup:
+        labels = [
+            "HPV+",
+            "HR Result",
+            "HR Cytology",
+            "HR Histology",
+            "HPV16+",
+            "HPV18+",
+        ]
```
[the remark this answers is not in the document]
Comment (anders0204) [reply bokeh-demo/examples/landing_page/main.py]: lovely

File: bokeh-demo/examples/landing_page/main.py [hunk 7]
```diff
@@ -213,28 +209,30 @@ def min_n_screenings_slider_pids(self) -> list[int]:
...
-            warnings.warn(str(e))
-            print("Falling back to .csv loading. This will affect performance.")
+            logger.exception(e)
+            logger.warning(
+                "Falling back to .csv loading. This will affect performance."
+            )
             data_manager = DataManager.read_from_csv(
                 settings.data_paths.screening_data_path,
                 settings.data_paths.dob_data_path,
+                read_hpv=True,
```
Comment (Northo) [inline bokeh-demo/examples/landing_page/main.py:235]: Nice catch!

**Review by Northo — CHANGES_REQUESTED** [review]
Very happy with the changes you made!

Three points I'd like addressing, then it's good to merge!

- https://github.com/Simula-Consulting/decipher/pull/83#discussion_r1243262589
Is the int conversion needed? If so: there should not be any reason for these to return non-ints. Is there a bug somewhere?

- https://github.com/Simula-Consulting/decipher/pull/83#discussion_r1243230457
I think this name should be changed.

- https://github.com/Simula-Consulting/decipher/pull/83#discussion_r1243181457
Is this a TODO? I don't understand what the comment means.

File: bokeh-demo/examples/landing_page/main.py [hunk 8]
```diff
@@ -1,25 +1,249 @@
...
+            map(
+                int,
+                list(
+                    set(age_pids)
+                    & set(button_pids)
+                    & set(min_age_last_exam_pids)
+                    & set(min_n_screenings_pids)
+                ),
+            )
+        )
```
[the remark this answers is not in the document]
Comment (anders0204) [reply bokeh-demo/examples/landing_page/main.py]: They are numpy ints, and they had to be int-ints

File: bokeh-demo/examples/landing_page/main.py [hunk 9]
```diff
@@ -1,25 +1,249 @@
...
+                )
+            ].index
+        )
+        return age_pids
+
+    def checkbox_button_pids(self) -> list[int]:
+        """Function to get the pids that have the selected attributes"""
+
+        def _get_hpv_pids(hpv_type: str) -> list[int]:
+            """Function to get the pids that have the selected HPV type"""
```
[the remark this answers is not in the document]
Comment (anders0204) [reply bokeh-demo/examples/landing_page/main.py:121]: No this is strictly for the hpv results? I filter to only look at at HPV positive results

File: bokeh-demo/examples/landing_page/main.py [hunk 10]
```diff
@@ -1,25 +1,249 @@
...
+        try:
+            data_manager = DataManager.from_parquet(settings.data_paths.base_path)
+        except (FileNotFoundError, ImportError, ValueError) as e:
+            warnings.warn(str(e))
+            print("Falling back to .csv loading. This will affect performance.")
+            data_manager = DataManager.read_from_csv(
+                settings.data_paths.screening_data_path,
+                settings.data_paths.dob_data_path,
+            )
+        # Add column for age in years
```
[the remark this answers is not in the document]
Comment (anders0204) [reply bokeh-demo/examples/landing_page/main.py:237]: oops

**Pull request merged**